docs/configuration.md
### Configuration

#### Example Configuration File

```js
{
    "app": {
        "name": "Project Name Here"
    },
    "server": {
        "host": "127.0.0.1",
        "port": 3020,
        "socketTimeoutSec": 30
    },
    "api": {
        "host": "127.0.0.1",
        "port": 3000
    },
    "auth": {
        "tokenUrl":"/token",
        "clientId":"testClient",
        "secret":"superSecret"
    },
    "aws": {
      "accessKeyId": "<your key here>",
      "secretAccessKey": "<your secret here>",
      "region": "eu-west-1"
    },
    "caching": {
        "ttl": 300,
        "directory": {
            "enabled": true,
            "path": "./cache/web/",
            "extension": "html"
        },
        "redis": {
            "enabled": false,
            "host": "project.stimkl.oh.0001.euw1.cache.amazonaws.com",
            "port": 6379
        }
    },
    "dust": {
        "cache": true,
        "debug": false,
        "debugLevel": "INFO",
        "whitespace": false
    },
    "headers": {
        "useGzipCompression": true,
        "cacheControl": {
          "text/css": "public, max-age=86400"
        }
    },
    "logging": {
        "enabled": true,
        "path": "./log",
        "filename": "web",
        "extension": "log",
        "accessLog": {
            "enabled": true,
            "fileRotationPeriod": "1d",
            "fileRetentionCount": 14,
            "kinesisStream": "dadi_web_test"
        }
    },
    "rewrites": {
        "datasource": "rewrites",
        "path": "workspace/routes/rewrites.txt",
        "forceTrailingSlash": true
    },
    "global" : {
        "baseUrl": "http://www.example.com"
    },
    "debug": true,
    "allowJsonView": true
}
```

#### Property Description

Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
debug           | If true, enables a debug panel on every page containing the loaded data and execution stats   |     false          | true
allowJsonView           |  If true, allows ?json=true in the querystring to return a view of the raw data loaded for the page  |   false            | true

###### Section: `server`

 Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
host           |    |               | api.example.com       
port           |    |               | 3000
socketTimeoutSec           | The number of seconds to wait before closing an idle socket   |        30       | 10

###### Section: `api`

 Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
host           |    |               | api.example.com       
port           |    |               | 3000

###### Section: `auth`

 Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
tokenUrl           |    |               |     "/token"   
clientId           |    |               |        "test123"
secret           |    |               |         "superSecret"

###### Section: `caching`

 Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
ttl           |    |               |  300      
directory           | Configuration block for caching using a local filesystem directory   |               |
directory.enabled           | If true, cache files will be stored on disk using the settings below. Either directory or redis caching must be enabled for caching to work.   | true              | true
directory.path           | The directory to use for storing cache files, relative to the root of the application   |    "./cache/web"           |  "./cache/web"
directory.extension           | The file extension to use for cache files   |    "html"           |  "html"
redis           | Configuration block for caching using a Redis caching service   |               |
redis.enabled           | If true, cache files will be stored in the Redis cache store using the settings below. Either directory or redis caching must be enabled for caching to work.   | false              | true
redis.host           | The host for the Redis caching service   |    ""           |  See example config above.
redis.port           | The port for the Redis caching service   |    6379           |  6379


###### Section: `dust`

 Property       | Description         | Default value  |  Example
:---------------|:--------------------|:---------------|:--------------
cache           |    									|       true     | true       
debug           |    									|       true     | true       
debugLevel      |                     |       "DEBUG"  | "DEBUG"
whitespace      |                     |       true     | false

###### Section: `headers`

 Property       | Description         | Default value  |  Example
:---------------|:--------------------|:---------------|:--------------
useGzipCompressiom           |    									|       true     | true       
cacheControl           |    									|       true     | true       

###### Section: `logging`

 Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
enabled           | If true, logging is enabled using the following settings.   |            true   | true       
path           | The absolute or relative path to the directory for log files.   |       "./log"        | "/data/app/log"
filename           | The filename to use for the log files. The name you choose will be given a suffix indicating the current application environment. |    "web"           | "your_application_name"      
extension           | The extension to use for the log files.  |  "log"   | "txt"

###### Section: `logging.accessLog`

Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
enabled           | If true, HTTP access logging is enabled. The log file name is similar to the setting used for normal logging, with the addition of 'access'. For example `web.access.log`.   |            true   | true       
fileRotationPeriod           | The period at which to rotate the access log file. This is a string of the format '$number$scope' where '$scope' is one of 'ms', 'h', 'd', 'w', 'm', 'y'. The following names can be used 'hourly' (= '1h'), 'daily (= '1d'), 'weekly' ('1w'), 'monthly' ('1m'), 'yearly' ('1y').   |       "1d"        | "daily"
fileRetentionCount           | The number of rotated log files to keep. |    7           | 14
kinesisStream           | An AWS Kinesis stream to write to log records to. |  ""   | "dadi_web_aws_kinesis"

###### Section: `logging.aws`

Property       | Description                 | Default value  |  Example
:---------------|:----------------------------|:---------------|:--------------
accessKeyId           |    |    ""   | ""
secretAccessKey           |    |       ""        | ""
region           |  |    ""           | ""

###### Section: `rewrites`

Property           | Description                 | Default value  |  Example
:------------------|:----------------------------|:---------------|:--------------
datasource         |            |  ""   |  "rewrites"
path               |            | ""    | "workspace/routes/rewrites.txt"
forceTrailingSlash |            | false | true

###### Section: `global`

The `global` section can be used for any application parameters that should be available for use in page templates, such as asset locations, 3rd party account identifiers, etc

```js
"global" : {
  "baseUrl": "http://www.example.com"
}
```

In the above example `baseUrl` would be availabe to a page template and could be used in the following way:

```html
<html>
<body>
  <h1>Welcome to DADI Web</h1>
  <img src="{baseUrl}/images/welcome.png"/>
</body>
</html>
```


HISTORY.md
0.1.7 / 2015-12-06
===================
* Config:
  - Add config option for socket timeout, defaults to 30 seconds

* Keepalive header added to Serama data & auth requests



0.1.7 / 2015-11-26
===================

* Server:
  - Error if the configured API can't be reached
  - Simplify middleware loading
* Logging:
  - Simplify error logging
  - Provide configuration for logging to a Sentry server
  - Remove Slack logging option, as this can be done from Sentry
* Config:
  - Provide configuration for logging to a Sentry server

0.1.6 / 2015-11-12
===================

  * Cache:
    - Don't cache requests that use ?json=true in the querystring
    - Provides better integration with Express patterns
  * Debug:
    - When debug: true is set in config, a debug panel is added to the right side of the page with
      a view of the loaded data and various execution stats. Note this won't work if your data object contains
      Javascript ads with no CDATA declaration.
  * Logging:
    - Locate client IP in request headers when behind a load balancer
    - Fix logging in default error handler
    - Addition of AWS Kinesis and Slack logging options	(see example files or config.js)
    - Don't log to slack in test mode
  * Config:
    - Add support for serving content from virtual directories (see example files or config.js)
    - Add support for Gzip compression (see example files or config.js)
    - Add support for configurable cache control headers (see example files or config.js)
    - Add application name (see example files or config.js)
  * Pages:
    - Extend `toPath` method to cover all routes in a page
    - Addition of new getComponent method on the Server to return a page by key
  * Datasources:
    - Remove json param from query before building filter
  * Tests
    - improve test suite coverage
    - ensure test environment is set before tests run
  * Misc:
    - rename public to sample-public to avoid overwrite when updating
    - check page file extension when reloading to avoid processing swap files


config.js
var convict = require('convict');

// Define a schema
var conf = convict({
  app: {
    name: {
      doc: "The applicaton name",
      format: String,
      default: "DADI Web (Repo Default)"
    }
  },
	server: {
    host: {
      doc: "The IP address the web application will run on",
      format: '*',
      default: '0.0.0.0'
    },
    port: {
      doc: "The port the web application will bind to",
      format: 'port',
      default: 8080
    },
    name: {
      doc: "The server name.",
      format: String,
      default: "DADI (Web)"
    },
    socketTimeoutSec: {
      doc: "The number of seconds to wait before closing an idle socket",
      format: Number,
      default: 120
    }
  },
	api: {
    host: {
      doc: "The IP address the DADI API application runs on",
      format: '*',
      default: '0.0.0.0'
    },
    port: {
      doc: "The port for the DADI API application",
      format: 'port',
      default: 8080
    },
    enabled: {
      doc: "Determines whether this web instance requires access to the API",
      format: Boolean,
      default: true
    }
  },
  auth: {
  	tokenUrl: {
      doc: "",
      format: String,
      default: "/token"
    },
    clientId: {
      doc: "",
      format: String,
      default: "testClient"
    },
    secret: {
      doc: "",
      format: String,
      default: "superSecret"
    }
  },
  aws: {
    accessKeyId: {
      doc: "",
      format: String,
      default: ""
    },
    secretAccessKey: {
      doc: "",
      format: String,
      default: ""
    },
    region: {
      doc: "",
      format: String,
      default: ""
    }
  },
  caching: {
    ttl: {
      doc: "",
      format: Number,
      default: 300
    },
    directory: {
      enabled: {
        doc: "If enabled, cache files will be saved to the filesystem",
        format: Boolean,
        default: true
      },
      path: {
        doc: "The relative path to the cache directory",
        format: String,
        default: "./cache/web"
      },
      extension: {
        doc: "The extension to use for cache files",
        format: String,
        default: "html"
      }
    },
    redis: {
      enabled: {
        doc: "If enabled, cache files will be saved to the specified Redis server",
        format: Boolean,
        default: false
      },
      host: {
        doc: "The Redis server host",
        format: String,
        default: "tresting.qvhlji.ng.0001.euw1.cache.amazonaws.com"
      },
      port: {
        doc: "The port for the Redis server",
        format: 'port',
        default: 6379
      },
      password: {
        doc: "",
        format: String,
        default: ""
      }
    }
  },
  dust: {
  	cache: {
      doc: "Determines if Dust caching is enabled",
      format: Boolean,
      default: true
    },
    debug: {
      doc: "",
      format: Boolean,
      default: true
    },
    debugLevel: {
      doc: "",
      format: String,
      default: "DEBUG"
    },
    whitespace: {
      doc: "",
      format: Boolean,
      default: false
    }
  },
  headers: {
    useGzipCompression: {
      doc: "If true, uses gzip compression and adds a 'Content-Encoding:gzip' header to the response.",
      format: Boolean,
      default: false
    },
    cacheControl: {
      doc: "A set of custom cache control headers for different content types. For example 'cacheControl': { 'text/css': 'public, max-age=1000' }",
      format: Object,
      default: {
        "text/css": "public, max-age=86400",
        "text/javascript": "public, max-age=86400",
        "application/javascript": "public, max-age=86400"
      }
    }
  },
  logging: {
  	enabled: {
      doc: "If true, logging is enabled using the following settings.",
      format: Boolean,
      default: true
    },
    path: {
      doc: "The absolute or relative path to the directory for log files.",
      format: String,
      default: "./log"
    },
    filename: {
      doc: "The name to use for the log file, without extension.",
      format: String,
      default: "web"
    },
    extension: {
      doc: "The extension to use for the log file.",
      format: String,
      default: "log"
    },
    accessLog: {
      enabled: {
        doc: "If true, HTTP access logging is enabled. The log file name is similar to the setting used for normal logging, with the addition of 'access'. For example `web.access.log`.",
        format: Boolean,
        default: true
      },
      fileRotationPeriod: {
        doc: "The period at which to rotate the access log file. This is a string of the format '$number$scope' where '$scope' is one of 'ms' (milliseconds), 'h' (hours), 'd' (days), 'w' (weeks), 'm' (months), 'y' (years). The following names can be used 'hourly' (= '1h'), 'daily (= '1d'), 'weekly' ('1w'), 'monthly' ('1m'), 'yearly' ('1y').",
        format: String,
        default: "1d"  // daily rotation
      },
      fileRetentionCount: {
        doc: "The number of rotated log files to keep.",
        format: Number,
        default: 7    // keep 7 back copies
      },
      kinesisStream: {
        doc: "An AWS Kinesis stream to write to log records to.",
        format: String,
        default: ""
      }
    },
    sentry: {
      enabled: {
        doc: "If true, error logging to a Sentry server is enabled.",
        format: Boolean,
        default: false
      },
      dsn: {
        doc: "The 'DSN' to use for Sentry logging. It should be similar to 'https://693ef18da3184cffa82144fde2979cbc:a0651b0286784761a62ef8e8fc128722@app.getsentry.com/59524'.",
        format: String,
        default: ""
      }
    }
  },
  rewrites: {
    datasource: {
      doc: "",
      format: String,
      default: ""
    },
    path: {
      doc: "",
      format: String,
      default: ""
    },
    forceTrailingSlash: {
      doc: "",
      format: Boolean,
      default: false
    }
  },
  env: {
    doc: "The applicaton environment.",
    format: ["production", "development", "test", "qa"],
    default: "development",
    env: "NODE_ENV",
    arg: "node_env"
  },
  virtualDirectories: {
    doc: "Allows specifying folders relative to the root of the application where additional static content may reside. An array entry should like look { path: 'data/legacy_features', index: 'default.html', forceTrailingSlash: false } ",
    format: Array,
    default: []
  },
  debug: {
    doc: "If true, debug mode is enabled and a panel containing the JSON loaded for each page is displayed alongside the normal content.",
    format: Boolean,
    default: false
  },
  allowJsonView: {
    doc: "If true, allows appending ?json=true to the querystring to view the raw JSON output for each page.",
    format: Boolean,
    default: false
  }
});

// Load environment dependent configuration
var env = conf.get('env');
conf.loadFile('./config/config.' + env + '.json');

// Perform validation
conf.validate({strict: false});

module.exports = conf;
